### PR TITLE
Add REFETCH and IGNORE rangeBehaviors

### PR DESCRIPTION
**Pull request by xuorig**
As discussed in #574, #542

Introduce 2 new range behaviors:
  - `REFETCH`: Will refetch the entire connection (to squelch the warning when no `rangeBehavior` matches the tracked connection)
  - `IGNORE`: Replaces using null, means the range should not be refetched at all.

I've deprecated `null`, maybe we don't need to, what do you think ?

@yungsters @steveluscher 

let me what you think! 🍻

### CODE REVIEW

File: src/traversal/writeRelayUpdatePayload.js [hunk 1]
```diff
@@ -413,7 +413,14 @@ function addRangeNode(
     null;
 
   // no range behavior specified for this combination of filter calls
-  if (!rangeBehavior) {
+  if (!rangeBehavior || rangeBehavior === IGNORE) {
+    if (!rangeBehavior) {
+      warning(
+        false,
```
Comment (wincent) [inline src/traversal/writeRelayUpdatePayload.js]: What do you think of?

```
warning(
  !rangeBehaviour,
  ...
);
```

instead of the `if`?

File: src/mutation/__tests__/RelayMutationQuery-test.js [hunk 2]
```diff
@@ -305,6 +305,96 @@ describe('RelayMutationQuery', () => {
...
+              }
+            }
+          }
+        }
+      `);
+      expect(node)
+        .toEqualQueryNode(expected);
+    });
+
+    it('Range is not refetched at all when rangeBehavior is set to IGNORE', () => {
```
Comment (wincent) [inline src/mutation/__tests__/RelayMutationQuery-test.js]: (nit) Lower case "r" in "Range"

File: src/mutation/RelayMutationQuery.js [hunk 3]
```diff
@@ -210,26 +212,32 @@ const RelayMutationQuery = {
...
-            'and to enable optimistic mutations.',
-            trackedConnection.getStorageKey(),
-            parentName,
-            parentID
-          );
+          // If the connection is not in `rangeBehaviors` or we have explicitly
+          // set the behavior to `refetch`, re-fetch it.
+          if (rangeBehaviorValue !== REFETCH) {
+            warning(
+              false,
```
Comment (wincent) [inline src/mutation/RelayMutationQuery.js]:
```
warning(
   rangeBehavior === REFETCH,
  ...
);
```

?

It's possible we're doing the `warning(false, ...)` thing all over the place to avoid the overhead of a called to a function, but this is not a hot code path and outside of `__DEV__` the function is just `emptyFunction`, so close to free...

Comment (xuorig) [inline src/mutation/RelayMutationQuery.js]: yup that makes a lot of sense! Will change it.

File: src/legacy/mutation/GraphQLMutatorConstants.js [hunk 4]
```diff
@@ -16,6 +16,8 @@ const GraphQLMutatorConstants = {
   APPEND: 'append',
   PREPEND: 'prepend',
   REMOVE: 'remove',
+  IGNORE: 'ignore',
+  REFETCH: 'refetch',
```
Comment (wincent) [inline src/legacy/mutation/GraphQLMutatorConstants.js]: (nit) Should probably alpha order these.

**Comment by xuorig** [comment]
Just realised failing tests, fixing this.

File: src/mutation/RelayMutationQuery.js [hunk 5]
```diff
@@ -23,6 +23,8 @@ const RelayQuery = require('RelayQuery');
 import type RelayQueryTracker from 'RelayQueryTracker';
 const RelayRecord = require('RelayRecord');
 import type {Variables} from 'RelayTypes';
+const GraphQLMutatorConstants = require('GraphQLMutatorConstants');
+const REFETCH = GraphQLMutatorConstants.REFETCH;
```
Comment (wincent) [inline src/mutation/RelayMutationQuery.js]: Could do:

    const {REFETCH} = GraphQLMutatorConstants;

of even collapse the two lines:

    const {REFETCH} = require('GraphQLMutatorConstants');

File: src/mutation/__tests__/RelayMutationQuery-test.js [hunk 6]
```diff
@@ -305,6 +305,96 @@ describe('RelayMutationQuery', () => {
       };
     });
 
+    it('refetches the whole range when the matching rangeBehavior is set to REFETCH', () => {
```
Comment (wincent) [inline src/mutation/__tests__/RelayMutationQuery-test.js]: Our internal linter is probably going to complain about this long line. Can you rephrase it as something like:

    it('refetches the whole range when the rangeBehavior is REFETCH', () => {

**Comment by wincent** [comment]
Looks pretty good @xuorig. Thanks for this!

I'll let it sit here for couple of days to give others a chance to comment before importing.

**Comment by xuorig** [comment]
thanks for the quick review @wincent! I addressed your comments.

**Comment by xuorig** [comment]
@wincent anything I can do to see this merged :) ?

**Comment by wincent** [comment]
I'll pull it into our internal CR system and get the ball rolling.

**Comment by wincent** [comment]
@facebook-github-bot import

**Comment by xuorig** [comment]
I don't know if that worked @wincent :panda_face:

**Comment by wincent** [comment]
That bot... I've used it to import at least a couple things today. This time I'm just going to import it manually.

**Comment by wincent** [comment]
Turns out it was an issue with Unicode in your commit (your name). I believe the import will work now, so giving it another try.

**Comment by wincent** [comment]
@facebook-github-bot import

**Comment by xuorig** [comment]
Blame my parents for naming me this way :sob:

**Comment by wincent** [comment]
After a long saga, the import worked, code review happened, I fixed some lint warnings and nits for you and it landed. Thanks once again for contributing!

**Comment by xuorig** [comment]
:tada:
